### DISCUSSION TITLE
Can I return multipe error types in transaction?

### DISCUSSION THREAD

**Discussion by knota-lab** [diesel-rs/diesel#2872] [Q&A]
I want to custom error in `transaction`. And I have been implemented `From<diesel::result::Error>`. But it seems like only one type error could be return. Is there any way to return multiple different error types?
![image](https://user-images.githubusercontent.com/26835087/130456844-70e3cf4a-094e-4f8c-8a44-808a15e37733.png)

**Comment by weiznich** [comment]
Please don't post code as picture as this is not searchable by anyone. Generally speaking any post here should contain the following information:
* All relevant code (including imports, `table!` definitions, definitions of referenced types)
* A list of things that you've already tried + error messages how this did not work
* Exact versions of diesel, rustc and information about which backend and operating system where used.

**Reply by knota-lab** [reply]
I streamlined this part of the code. 
```rust
#[cfg(test)]
mod diesel_test {
    use diesel::result::Error;
    use diesel::{Connection, MysqlConnection};

    #[derive(Debug)]
    /// 版本不一致错误
    struct VersionInconsistencyError {
        pub message: String,
        pub code: String,
    }
    
    impl From<Error> for VersionInconsistencyError {
        fn from(_: Error) -> Self {
            VersionInconsistencyError {
                message: "版本不一致".to_string(),
                code: "VersionInconsistencyError".to_string(),
            }
        }
    }    

    fn establish_connection() -> MysqlConnection {
        let url = ::std::env::var("DATABASE_URL").unwrap();
        MysqlConnection::establish(&url).unwrap()
    }

    fn multiple_error(i: i32) {
        let conn = &mut establish_connection();
        conn.transaction(|| {
            if i == 1 {
                return Err(Error::NotFound);
            }

            if i == 2 {
                return Err(VersionInconsistencyError::from(Error::NotFound));
            }
            
            Ok(i)
        }).unwrap();
    }

    #[test]
    pub fn transaction_test() {
        multiple_error(0);
    }
}
```

It got this error.
```
   Compiling document-manager v0.1.0 (H:\rust-workspace\document-manager)
error[E0308]: mismatched types
  --> src\test\diesel_test.rs:35:28
   |
35 |                 return Err(VersionInconsistencyError::from(Error::NotFound));
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `diesel::result::Error`, found struct `test::diesel_test::diesel_test::VersionInconsistencyError`
   |
note: return type inferred to be `diesel::result::Error` here
  --> src\test\diesel_test.rs:31:24
   |
31 |                 return Err(Error::NotFound);
   |                        ^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `document-manager
```

Operating system is `win10`. Rustc is `rustc 1.54.0-nightly (657bc0188 2021-05-31)`. The diesel version is this.
```
diesel = { version = "1.3", features = ["mysql", "r2d2", "chrono"] }
diesel_migrations = "1.3"
```

**Reply by weiznich** [reply]
As already mentioned in https://github.com/diesel-rs/diesel/discussions/2859 diesel requires that you return a single error type from the closure passed to `transaction` (as this is the case for any other rust function es well). Have a look a [enums](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html) for composite error types.
